### PR TITLE
Properly infer intervals of pcolormesh when plotting on logscale

### LINKED ISSUE #5333: Unexpected behaviour of pcolormesh with logscale in x or y axes
**What happened**:

When using `xarray.DataArray.plot.pcolormesh` with `yscale="log"` the output figure distorts the y axis and thus doesn't plot the data correctly. I think the problem is related to the inference of the intervals through the `xarray.plot.utils._infer_intervals_breaks()` function.

**What you expected to happen**:

I expect the `yscale="log"` argument to correctly display the y axis in a logscale while precisley plotting the data. In my opinion, the inference of the intervals for the `pcolormesh` should be also done in log scales.

**Minimal Complete Verifiable Example**:

```python
import numpy as np
import xarray as xr
import matplotlib.pyplot as plt

x = np.linspace(-3, 4, 15)
y = np.logspace(-4, 3, 8)

xx, yy = np.meshgrid(x, y)
z = np.exp(-(xx ** 2) - np.log10(yy) ** 2) - np.exp(
    -((xx - 1) ** 2) - (np.log10(yy) - 1) ** 2
)

da = xr.DataArray(
    z,
    coords={"x": x, "y": y},
    dims=("y", "x"),
)

da.plot(yscale="log")
plt.show()

```

Output:

Figure_1

**Anything else we need to know?**:

I pass `infer_intervals=False` I get a plot that looks more to the expected one:

```python
import numpy as np
import xarray as xr
import matplotlib.pyplot as plt

x = np.linspace(-3, 4, 15)
y = np.logspace(-4, 3, 8)

xx, yy = np.meshgrid(x, y)
z = np.exp(-(xx ** 2) - np.log10(yy) ** 2) - np.exp(
    -((xx - 1) ** 2) - (np.log10(yy) - 1) ** 2
)

da = xr.DataArray(
    z,
    coords={"x": x, "y": y},
    dims=("y", "x"),
)

da.plot(yscale="log", infer_intervals=False)
plt.show()

```

Figure_3

The problem with deactivating the `infer_intervals` is that matplotlib crops the last row and column of the 2d grid, because we would be selecting`shading="flat"`, getting the following deprecation warning from matplotlib:
```
/home/santi/git/xarray/xarray/plot/plot.py:1033: MatplotlibDeprecationWarning:
shading='flat' when X and Y have the same dimensions as C is deprecated since
3.3.  Either specify the corners of the quadrilaterals with X and Y, or pass
shading='auto', 'nearest' or 'gouraud', or set rcParams['pcolor.shading'].
This will become an error two minor releases later. 
primitive = ax.pcolormesh(x, y, z, **kwargs)
```

That's why I think that the inference of the breaks should also be done in log scales.

**Environment**:

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: 49aa235fc63706dc145128dd13fdf139b9d5bc6e
python: 3.9.4 | packaged by conda-forge | (default, May 10 2021, 22:13:33) 
[GCC 9.3.0]
python-bits: 64
OS: Linux
OS-release: 5.12.1-2-MANJARO
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.utf8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.10.6
libnetcdf: 4.7.4

xarray: 0.18.1.dev33+g49aa235f
pandas: 1.2.4
numpy: 1.20.1
scipy: 1.6.3
netCDF4: 1.5.6
pydap: installed
h5netcdf: 0.11.0
h5py: 3.2.1
Nio: None
zarr: 2.8.1
cftime: 1.4.1
nc_time_axis: 1.2.0
PseudoNetCDF: installed
rasterio: 1.2.3
cfgrib: 0.9.9.0
iris: 2.4.0
bottleneck: 1.3.2
dask: 2021.05.0
distributed: 2021.05.0
matplotlib: 3.4.2
cartopy: 0.19.0.post1
seaborn: 0.11.1
numbagg: installed
pint: 0.17
setuptools: 49.6.0.post20210108
pip: 21.1.1
conda: None
pytest: 6.2.4
IPython: None
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by santisoler**
When inferring the interval breaks, the `coord` and first linearized if
`scale="log"` and then the intervals are brought back to the original scale
before being returned to `pcolormesh`. Raise an error if at least one of the
elements of the coodinates has a non-positive element when using a logscale.

- [x] Closes #5333
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by shoyer** [comment]
this is great, thank you!

could you please look into adding a unit-test, to ensure that this stays fixed?

see http://xarray.pydata.org/en/stable/contributing.html#writing-tests

**Comment by santisoler** [comment]
Thanks @shoyer!

> could you please look into adding a unit-test, to ensure that this stays fixed?

Yes! I got a little bit busier after I opened this PR and couldn't find the time to do so. Let me find a gap to add some tests. I'll ping you as soon as the PR is ready for a review. Thanks!

**Comment by santisoler** [comment]
@shoyer Now I think this PR is ready to be reviewed. I added one test function for the `_infer_interval_breaks` private function with the optional `scale="log"` argument. I've also added a new test class for checking if calling `pcolormesh` with both `xscale="log"` and `yscale="log"` works as expected. I added the bug fix to `whats-new.rst` and run `pre-commit run --all-files` locally with success. I don't think we need to change anything in `api.rst`.

Let me know if you think I should change or add anything else, and feel free to make any suggestion on the review.

**Comment by santisoler** [comment]
@shoyer I've just fixed some errors on the `TestAxesKwargs` tests that were using logscales without passing any coordinates.
I've also added an error raising if the passed coordinates have at least one non-positive element.
Hope CIs pass now. Let me know if you feel that something else should be changed.

File: doc/whats-new.rst [hunk 1]
```diff
@@ -45,6 +45,9 @@ Deprecations
 Bug fixes
 ~~~~~~~~~
 
+- Plotting a pcolormesh with ``xscale="log"`` and/or ``yscale="log"`` works as
+  expected after improving the way the interval breaks are generated (:issue:`5333`).
+  By `Santiago Soler <https://github.com/santisoler>`_
```
Comment (shoyer) [inline doc/whats-new.rst:50]: When you get the chance, please merge in the latest changes from git master and move this to the section for 0.18.3.

**Review by shoyer — APPROVED** [review]
This looks great, thank you!

File: xarray/tests/test_plot.py [hunk 2]
```diff
@@ -2638,6 +2717,16 @@ def test_ncaxis_notinstalled_line_plot(self):
     DataArray(easy_array((10, 3, 2))),
 ]
 
+# Define a few arrays for testing scale="log"
+test_da_list_log = [
```
Comment (max-sixty) [inline xarray/tests/test_plot.py:2721]: @santisoler please could we make this into a fixture? If it's in the base of the module, then it'll be run on every import of this file, which will slow down collecting tests (even those that don't use it)

**Review by max-sixty — CHANGES_REQUESTED** [review]

File: xarray/tests/test_plot.py (hunk 2 above)
Comment (santisoler) [reply xarray/tests/test_plot.py]: I've just transformed the `test_da_list_log` as a fixture method in `TestAxesKwargs` with the number of dimensions of the returned DataArray as a parameter of the fixture. I took the liberty to do the same with the preexisting `test_da_list `.

**Review by shoyer — APPROVED** [review]
LGTM

@max-sixty any remaining concerns with the test fixtures?

**Review by dcherian — APPROVED** [review]

**Comment by dcherian** [comment]
I think we can address any concerns about test fixtures in a future PR.

Thanks @santisoler. I see this is your first contribution. welcome to xarray!

**Comment by santisoler** [comment]
Thanks @dcherian! I really liked the PR process in xarray, having 3 reviewers here was awesome and a crucial part for building quality code. Thanks for your hard work and comments! Hope this to be the first PR or many more!

**Pull request merged**